### PR TITLE
fix: panic on invalid file:// module specifier

### LINKED ISSUE #8958: thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ()', runtime\permissions.rs:570:37
Error:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ()', runtime\permissions.rs:570:37
```
Step to reproduce:
```
const foo = import("file:///")
```
with RUST_BACKTRACE=full
```
Deno 1.6.3
exit using ctrl+d or close()
> const foo = import("file:///")
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ()', runtime\permissions.rs:570:37
stack backtrace:
   0:     0x7ff62eb509a6 - <unknown>
   1:     0x7ff62e61213b - <unknown>
   2:     0x7ff62eb4f6d8 - <unknown>
   3:     0x7ff62eb4ea53 - <unknown>
   4:     0x7ff62eb56505 - <unknown>
   5:     0x7ff62eb564c7 - <unknown>
   6:     0x7ff62eb5646f - <unknown>
   7:     0x7ff62e60f620 - <unknown>
   8:     0x7ff62e616a13 - <unknown>
   9:     0x7ff62e838e3b - <unknown>
  10:     0x7ff62e86f369 - <unknown>
  11:     0x7ff62e4e3b86 - <unknown>
  12:     0x7ff62e4ebb01 - <unknown>
  13:     0x7ff62e4e065e - <unknown>
  14:     0x7ff62e4fbcae - <unknown>
  15:     0x7ff62e4dc550 - <unknown>
  16:     0x7ff62e62ad08 - <unknown>
  17:     0x7ff62e63c462 - <unknown>
  18:     0x7ff62e4be34f - <unknown>
  19:     0x7ff62e4b5a98 - <unknown>
  20:     0x7ff62e584807 - <unknown>
  21:     0x7ff62e4b9165 - <unknown>
  22:     0x7ff62e79581a - <unknown>
  23:     0x7ff62e5b2ff5 - <unknown>
  24:     0x7ff62e363401 - <unknown>
  25:     0x7ff62e5c19ad - <unknown>
  26:     0x7ff62f81fd00 - CrashForExceptionInNonABICompliantCodeRange
  27:     0x7ff808377034 - BaseThreadInitThunk
  28:     0x7ff8088fd0d1 - RtlUserThreadStart
```
OS: Windows 10.0.19041.685
Deno: deno 1.6.3 (release, x86_64-pc-windows-msvc)
v8 8.8.294
typescript 4.1.3

This might be a Windows-specific issue, because when I try same line using WSL, I got this instead
```
Deno 1.6.3
exit using ctrl+d or close()
> const foo = import("file:///")
undefined
> foo
Promise { <rejected> TypeError: Is a directory (os error 21) }
```

### PR DESCRIPTION
**Pull request by Spoonbender**
Fixes #8958 
- `permissions.rs`: `check_specifier` no longer assumes a local-file specifier is valid
- `disk_cache.rs`: `get_cache_filename` gracefully handles invalid local-file specifiers

Working on tests now

### CODE REVIEW

**Comment by Spoonbender** [comment]
Added some tests.
I noticed some CI failures which are gone now, I'm not sure if I introduced some regression that causes a test to flicker, or if the test was flickering regardless. The test that failed is `local_sources_not_cached_in_memory`

File: cli/disk_cache.rs [hunk 1]
```diff
@@ -278,4 +281,28 @@ mod tests {
...
+      test_cases.push("file://");
+      test_cases.push("file:///");
+    }
+
+    for test_case in &test_cases {
+      let cache_filename =
+        cache.get_cache_filename(&Url::parse(test_case).unwrap());
+      assert_eq!(cache_filename, None);
+    }
+  }
```
Comment (ry) [inline cli/disk_cache.rs:307]: Since it's just one test case, remove this for loop

File: runtime/permissions.rs [hunk 2]
```diff
@@ -942,6 +947,24 @@ mod tests {
     }
   }
 
+  #[test]
+  fn check_invalid_specifiers() {
+    let perms = Permissions::allow_all();
+
+    let mut test_cases = vec![];
+
+    if cfg!(target_os = "windows") {
```
Comment (ry) [inline runtime/permissions.rs:956]: can this be removed? this test doesn't check anything on non-windows.

File: cli/disk_cache.rs (hunk 1 above)
Comment (Spoonbender) [reply cli/disk_cache.rs:307]: @ry not sure I follow? on Windows we push two additional test cases on lines 297-300

File: runtime/permissions.rs (hunk 2 above)
Comment (Spoonbender) [reply runtime/permissions.rs:956]: @ry added a non-Windows test case (a local-file URI which is invalid on UNIX)

**Review by ry — APPROVED** [review]
LGTM - thanks @Spoonbender

**Pull request merged**